[15.0][MIG+REF] sale_commission: Refactorization to split and extract the base

pedrobaeza opened the pull request:
Restarting from the excellent work done by Quartile, adding these extra things:

- Squash of the administrative commits.
- Make specific top-level menu and permissions for commissions.
- Don't depend on account for base module.
- Reorganize elements.
- Change some parts of the code to fit better ORM tools.
- Adjust permissions in children modules.

@Tecnativa TT36983

martiita committed:
sale_commissions: added module migrated to 6.1. (../../7.0-pexego_sale_commission rev 7)

Commit:
sale_commission. Ported to version 7.0 (../../7.0-pexego_sale_commission rev 15)

Commit:
Launchpad automatic translations update.

Commit:
[FIX+IMP] sale_commission: 4 things

* PEP 8 fixes
* Replace depreciated osv for orm and other depreciation fixes
* Code fix and optimization
* Add context

Commit:
onchange_partner_id() got an unexpected keyword argument 'context'

Commit:
#1180743 + missing context

tafaRU committed:
Removed wrong bracket from L197 of the settled.py

mileo committed:
Improve partnert view

gurneyalex committed:
adapt QA files to 8.0

gborelli committed:
[MIG] sale_commission: Migration to 8.0

* migrate models security and i18n
* add demo data and minor fixes
* add sale agent and commission views
* add stock dependency add views and product/stock overrides
* fixed related fields
* minor fixes: names and labels
* change 'onchange' attribute in view to api.onchange'
* improve comments about migration
* add invoice view
* add wizards
* use api.onchange
* add store attribute on res.partner.agent
* set default currency
* change state field to readonly
* migrate stock picking and settlements views
* fix invoice view
* fix invoice wizard
* change osv.except_osv to openerp.exceptions
* add missed report
* fix taxes in action_invoice_create

Commit:
<hr_commission><sale_commission><stock_commission>.

pedrobaeza committed:
[RFR] sale_commission: Total refactorization of the modules

* Take default commission when manual adding agents + error on settled filter
* Many2one pointer of comission section
* Fix translation problem + add commission total on sales orders
* Cancel settlement when the associated invoice is cancel and go back to invoiced when re-validated

pablocm-aserti committed:
sale.commission.settlement access error opening res.partner form for users w/o sale rights

Commit:
"sale" journal type on making agent invoice

On generate agent invoice you can only select journal of type "sale". This can cause the numbers of sales invoices not correlated.
IMHO, agent invoices must be of type purchase and this solves it, allowing to select journals of type purchase.

mbcosta committed:
Created new field to specify the commission creation according to the state of the invoice

oihane committed:
sale_commission: base amount for computation selectable You can now select if commission is calculated from subtotal or from margin

* added python tests

Commit:
Fix singleton error

pedrobaeza committed:
sale_commission: Refunds from invoices. Fixes #51

When creating a refund from an invoice, pass correct values
(converted as dictionary) for the agents lines, so that the refund
keep the same commissions (but reversed) than the original invoice.

ivantodorovich committed:
Change visibility of agents field to only show when partner is a customer. (instead of showing when agent is not an agent)

pedrobaeza committed:
sale_commission: Fix tests

Tests create invoices in a month previous the current date, which
makes that they fail on January, as previous fiscal year doesn't
exist. This is fixed creating invoices in the current date, and
calling the settle wizard with one month more. This won't fail
on December, because there's no need of the next fiscal year for
this operation

pedrobaeza committed:
sale_commission: Don't put supplier flag if not agent

Without this condition, all partners are suppliers, because the default
value for agent_type is agent and the onchange is executed.

ivantodorovich committed:
[IMP+FIX] sale_commission: 4 things:

* Show wizards results
* Fix refunds
* Fixes #61
* Fixes #63.

Commit:
sale_commission: : Migrated to 10.0

FIX after changing customer on SO with lines, can't save sale order anymore
See https://github.com/odoo/odoo/issues/17618
Also add button to recompute agents on SO lines

rgarnau committed:
add method to prepare agents

add same method for account invoice

eLBati committed:
FIX sale_commission quaterly settlement

as May would have June as period start:
((5 - 1) // 3 + 1) * 3

instead, the new formula:
>>> for m in range(1, 13):
...     print (m - 1) // 3 * 3 + 1
...
1
1
1
4
4
4
7
7
7
10
10
10

See
https://github.com/hurrinico/commission/pull/2
and
https://github.com/odoo/odoo/issues/17618

This extends the workaround, fixing "Record not found" after changing invoice date, for example

hendrixcosta committed:
Tradução pt_BR

etobella committed:
[MIG] sale_commission: Migrationt to 11.0

oihane committed:
[FIX] sale_commission: installation fails if en_US language is not installed

oca-transbot committed:
OCA Transbot updated translations from Transifex

rgarnau committed:
small fix

oca-transbot committed:
OCA Transbot updated translations from Transifex

JordiBForgeFlow committed:
sale_commision: temporary fix until https://github.com/odoo/odoo/issues/24702
 is fixed

etobella committed:
[IMP] Add hook on make invoice settlement

oca-transbot committed:
OCA Transbot updated translations from Transifex

lk-eska committed:
Make Section Model Accessible By Everyone

etobella committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (118 of 118 strings)

Translation: commission-11.0/commission-11.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-11-0/commission-11-0-sale_commission/es/

eLBati committed:
FIX sale_commission the following use case (#170)

- Go to Sales -> Customers
 - Filters -> Add Custom Filter: Agents contains test

Get

odoo/addons/base/res/res_partner.py", line 647, in name_search
    where_query = self._where_calc(args)
odoo/models.py", line 4038, in _where_calc
    domain = [('active', '=', 1)] + domain
TypeError: can only concatenate list (not "str") to list

Commit:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100,0% (118 of 118 strings)

Translation: commission-11.0/commission-11.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-11-0/commission-11-0-sale_commission/pt_BR/

aitorbouzas committed:
[IMP] sale_commission: Adding graph view and settlement report

pedrobaeza committed:
[IMP] sale_commission: Work on integration tests + improve invoice generation

* Delaying tests to when all modules are loaded, we avoid problems with required
  fields declared by other modules that have a default value, but can't still be loaded.
* Calling journal onchange for avoiding empty currency_id field

etobella committed:
[IMP] commission: Add two hooks on make settlements (#176)

rschnapka committed:
Translated using Weblate (German)

Currently translated at 96.6% (114 of 118 strings)

Translation: commission-11.0/commission-11.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-11-0/commission-11-0-sale_commission/de/

dcorio committed:
Translated using Weblate (Italian)

Currently translated at 89.8% (106 of 118 strings)

Translation: commission-11.0/commission-11.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-11-0/commission-11-0-sale_commission/it/

ljimenezsidoo committed:
FIX taxes in settlement invoice creation

javierjcf committed:
[FIX]sale_commission: When creating agent from agents fields, on partner view, default field agent must be true

etobella committed:
[IMP] sale_commission: Add hook on the create settlements (#189)

pedrobaeza committed:
[FIX+IMP] sale_commission: Reworked module

- Remove onchange hackings as they are not needed with new JS views.
- Extract common code between sales orders and invoices, creating 2 mixins.
- Adapt tests.
- Improve README with detailed explanations.
- README by fragments.
- Changed `settled` condition to a more suitable one, moving the condition to _skip_settlement.
- Improve interface for not removing editable option in lines: you have a new icon that popups
  a window with agents. This icon is available according certain conditions. Check README for
  more details.
- Add constraint for avoiding the modification of a settled commission.
- Add agent values when no other value is provided. This way, we add compatibility with
  modules like `sale_product_set` or `sale_order_variant_mgmt`.

oca-transbot committed:
[UPD] Update sale_commission.pot + Update translation files

Updated by Update PO files to match POT (msgmerge) hook in Weblate.

pedrobaeza committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (161 of 161 strings)

Translation: commission-11.0/commission-11.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-11-0/commission-11-0-sale_commission/es/

pedrobaeza committed:
[MIG] sale_commission: Migration to 12.0

* Standard procedure
* Add description to all models
* Adapt commission amount computation courtesy of Simone Rubi
* Adapt views
* Adapt code for UoM and dates
* Fixed dependencies in constraint
* Fixed BI view

ernestotejeda committed:
[IMP] sale_commission: Agent propagation + tests

dcorio committed:
[FIX] avoid KeyError

pedrobaeza committed:
[REF] sale_commission: Use SavepointCase

etobella committed:
[12.0][IMP] sale_commission: move check of the amount to the end (#204)

SimoRubi committed:
[FIX] sale_commission: Strictly check commission type

rdc02271 committed:
Added translation using Weblate (Portuguese (Portugal))

Currently translated at 70.1% (108 of 154 strings)

Translation: commission-12.0/commission-12.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-12-0/commission-12-0-sale_commission/pt_PT/

oca-travis committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: commission-12.0/commission-12.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-12-0/commission-12-0-sale_commission/

[UPD] Update sale_commission.pot

Commit:
[12.0][FIX] sale_commission: show agents field on partner form

eLBati committed:
Translated using Weblate (Italian)

Currently translated at 73.5% (114 of 155 strings)

Translation: commission-12.0/commission-12.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-12-0/commission-12-0-sale_commission/it/

marcelsavegnago committed:
Translated using Weblate (Portuguese (Brazil))

Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (155 of 155 strings)

Translation: commission-12.0/commission-12.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-12-0/commission-12-0-sale_commission/pt_BR/

pedrocs-exo committed:
Added translation using Weblate (Portuguese)

Translated using Weblate (Portuguese)

Currently translated at 97.4% (151 of 155 strings)

Translation: commission-12.0/commission-12.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-12-0/commission-12-0-sale_commission/pt/

pedrobaeza committed:
[FIX] sale_commission: Add view context by code

As Odoo hasn't got any mechanism for adding several elements to the context dict,
using XML view tools, we can have our context replaced by other module doing the
same operation.

Thus, we avoid the problem adding the context by code, as this context is vital
for having default agents commissions populated on the sales order lines.

liweijie0812 committed:
Added translation using Weblate (Chinese (Simplified))

Currently translated at 100.0% (155 of 155 strings)

Translation: commission-12.0/commission-12.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-12-0/commission-12-0-sale_commission/zh_CN/

thomaspot committed:
Added translation using Weblate (Dutch)

Currently translated at 98.1% (152 of 155 strings)

Translation: commission-12.0/commission-12.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-12-0/commission-12-0-sale_commission/nl_NL/

oca-travis committed:
[UPD] Update sale_commission.pot + Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: commission-12.0/commission-12.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-12-0/commission-12-0-sale_commission/

SimoRubi committed:
Translated using Weblate (Italian)

Currently translated at 75.0% (117 of 156 strings)

Translation: commission-12.0/commission-12.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-12-0/commission-12-0-sale_commission/it/

SimoRubi committed:
[IMP] sale_commission: Show signed subtotal

marcelsavegnago committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (157 of 157 strings)

Translation: commission-12.0/commission-12.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-12-0/commission-12-0-sale_commission/pt_BR/

pedrobaeza committed:
[FIX] sale_commission: No commission should be populated on vendor bills

Commit:
[IMP] sale_commission: Migration to 13.0 (partial)

pedrobaeza committed:
[MIG+IMP] sale_commission: Switch to computed writable + finish migration + other imps

* Computed writable saves a lot of code
* Fix incorrect migration code
* Fix tests
* Remove redundant or currently useless tests
* Adjust multi-company rules
* Rename fields following guidelines

Commit:
[IMP] : black, isort, prettier

pedrobaeza committed:
[IMP] sale_commission: Multi-company record rules

If not put, settling by a regular user with multi-company agents will have an
access error, and you will see settlements for all companies.

ibuioli committed:
Added translation using Weblate (Spanish (Argentina))

Currently translated at 100.0% (155 of 155 strings)

Translation: commission-13.0/commission-13.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-13-0/commission-13-0-sale_commission/es_AR/

pedrobaeza committed:
[FIX] sale_commission: Settlement report printing

Fixes #254

dennybiasiolli committed:
[13.0][FIX] sale_commission: fixing inheritance of view_res_partner_filter

from `base.view_res_partner_filter` to `account.res_partner_view_search`

`<filter name="supplier">` was removed in https://github.com/odoo/odoo/commit/3e97cff7797c4f7e695d20def6b920561b6a6525

sergio-teruel committed:
[IMP] sale_commission: Remove required attribute in settlement field. Added to view

When updating a table, modules that don't have the `required=True` setting of a field in their scope issue an `ALTER TABLE` at least twice per module. This blocks the entire table.

This module was doing that, and since the `res.partner` model is so ubiquitous, this was blocking concurrent updates.

By moving the required attribute to the views, the UX is the same, but concurrent updates work fine again.

@Tecnativa TT27096

Commit:
[MIG] sale_commission: Migration to 14.0

ao-landoo committed:
Core context is not removed from invoice_line_ids

joao-p-marques committed:
[FIX] sale_commission: Fix tests

"ref" field is required for supplier invoices. We should create the object correctly in the tests.

TT28236

victoralmau committed:
[IMP] sale_commission: Add optional in commission_status field related to order view

anasuarez1 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (155 of 155 strings)

Translation: commission-13.0/commission-13.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-13-0/commission-13-0-sale_commission/es/

SilvioGregorini committed:
FIX sale_commission: fix wrong constrain logic on settlement lines

renatonlima committed:
[FIX] required only if partner is agent

fredzamoabg committed:
[IMP] sale_commission: black, isort, prettier

Yvesldff committed:
Added translation using Weblate (French)

Currently translated at 92.9% (144 of 155 strings)

Translation: commission-14.0/commission-14.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-14-0/commission-14-0-sale_commission/fr/

Commit:
[14.0][IMP] sale_commission: Only show commission for customers

eLBati committed:
FIX sale_commission preventing to cancel invoice with settled commission lines.

Steps to reproduce:

 - Create an agent
 - Create and validate an invoice setting that agent
 - Settle the commission
 - Cancel the invoice

Observed: the invoice is canceled

Desired: cancellation is blocked

etobella committed:
[IMP] sale_commission: Allow to group invoices

max3903 committed:
[FIX] sale_commission

Zar21 committed:
[14.0][FIX] sale_commission: amount compute

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (156 of 156 strings)

Translation: commission-14.0/commission-14.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-14-0/commission-14-0-sale_commission/it/

primes2h committed:
Translated using Weblate (Italian)

Currently translated at 97.4% (152 of 156 strings)

Translation: commission-14.0/commission-14.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-14-0/commission-14-0-sale_commission/it/

Commit:
Translated using Weblate (Italian)

Currently translated at 98.7% (154 of 156 strings)

Translation: commission-14.0/commission-14.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-14-0/commission-14-0-sale_commission/it/

primes2h committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (156 of 156 strings)

Translation: commission-14.0/commission-14.0-sale_commission
Translate-URL: https://translation.odoo-community.org/projects/commission-14-0/commission-14-0-sale_commission/it/

antoniospneto committed:
[IMP] sale_commission allowing to filter sale orders and invoices by agents

felipemotter committed:
[FIX] invoice_date

pedrobaeza committed:
[IMP] sale_commission: New bi-weekly settlement period

There will be 2 settlement per month, one covering the first 15 days,
and the other for the rest of the month.

TT35385

olgamarcocb committed:
[IMP] sale_commission: Adapt Forward-port to 14

oihane committed:
[FIX] sale_commission: compute amount after changing commission, fixes #355

ilyasProgrammer committed:
[14.0][IMP] sale_commission: show Group Invoices label + hook for sale_commission_queued

AungKoKoLin1997 committed:
[IMP] sale_commission: black, isort, prettier

AungKoKoLin1997 committed:
[MIG] sale_commission: Migration to 15.0

pedrobaeza committed:
[REF] *commission*: Abstract even more the base module

- Make specific top-level menu and permissions for commissions.
- Don't depend on account.
- Reorganize elements.
- Change some parts of the code to fit better ORM tools.
- Adjust permissions in children modules.
- Add <sheet> to some form views to get modern view.

pedrobaeza commented:
@yostashiro @AungKoKoLin1997 the last steps I talked about last Friday are done and you can see the result here. Tell me if it's OK for you for finally merging this long awaited refactorization.

/ocabot migration sale_commission

yostashiro commented:
@pedrobaeza Thanks for the PR.  I was thinking that vendor bill creation part would be separated from `account_commission` to achieve finer granularity, but I think we are fine with the proposed changes.  We will try and review the details in the coming days.  Thanks again.

pedrobaeza commented:
Yeah, I thought about that also, but being not too much lines of code, and complying with the dependencies, I let it inside the same one.

pedrobaeza commented:
@francesco-ooops your review is welcome here.

francesco-ooops commented:
@pedrobaeza on the functional side, is there something new specific to test?

pedrobaeza commented:
The new functional scope is:

- Separate application for commissions.
- Manual settlements.
- Limited access from the other applications to commissions.
- Settlement wizard by type.

francesco-ooops reviewed (approved):
functionally LGTM,

since this is the base module I guess users would appreciate having settlement type selected by default

![image](https://user-images.githubusercontent.com/55489873/203358212-07d731db-97c5-4efa-aced-1410a0d1a5c8.png)

and a server action for settlement invoicing

![image](https://user-images.githubusercontent.com/55489873/203373496-3fbc6b60-dbde-4d75-b701-fa0d748837b5.png)

but these are just ux improvements

AungKoKoLin1997 reviewed (approved):
Functionally LGTM
In my opinion, I think enterprise users may confuse a little bit for home app screen UX. Two apps related with accounting will be displayed in home screen. But just a minor thing.

![Screenshot from 2022-11-22 09-35-20](https://user-images.githubusercontent.com/45355704/203481976-10886cc6-012e-40fb-b81a-2d0dc12ef9b2.png)


![Screenshot from 2022-11-22 09-35-40](https://user-images.githubusercontent.com/45355704/203482017-c70b060f-c97a-4038-a93a-ed52ed12a667.png)

yostashiro commented on commission/models/commission.py:
```diff
@@ -0,0 +1,54 @@
+        required=True,
+        default="gross_amount",
+    )
+
```
Thought we might want `settlement_type` field here to control the source-document specific requirement - for example, `invoice_state` should only be required when `settlement_type` is `sale_invoice`.  Currently `invoice_state` is required (in `account_commisson`) even through it is not used in case there are cases in the future where settlements are done from different models than invoice.

pedrobaeza replied:
I was modifying this, but then realize that this field is for controlling about which price is referred to, and this is something universal.

Anyway, I'm adding the settlement_type for being able to restrict a commission to a certain type of operation, or leave it blank for allowing in all of them.

pedrobaeza replied:
Already pushed in 433a3a2a8a0fdde46395e4e5dfc56f4da232d8af this change for being reviewed while I do the rest of the changes.

yostashiro commented on account_commission/models/res_partner.py:
```diff
@@ -0,0 +1,17 @@
+        comodel_name="commission.settlement",
+        inverse_name="agent_id",
+        readonly=True,
+    )
```
I think this is an oversight of our work, but should we move this back to `commission`? I don't think this should be in this module.

yostashiro commented on commission/models/commission_settlement.py:
```diff
@@ -0,0 +1,139 @@
+    )
+    can_edit = fields.Boolean(
+        compute="_compute_can_edit",
+        help="Technical field for determining if user can" " edit settlements",
```
```suggestion
        help="Technical field for determining if user can edit settlements",
```

yostashiro commented on commission/wizards/commission_make_settle.py:
```diff
@@ -0,0 +1,152 @@
+        help="Technical field for improving UX when no extra *commission is installed.",
+    )
+
+    def _compute_can_settle(self):
```
```suggestion
    @api.depends("date_to")
    def _compute_can_settle(self):
```
Wasn't totally sure if this suggestion would be valid, but compute doesn't seem to run without a trigger.  Here is an image from runboat.
![image](https://user-images.githubusercontent.com/7766939/203697965-e557b305-905b-49dc-8d79-c8ee47045315.png)

yostashiro commented on account_commission/views/account_commission_settlement_view.xml:
```diff
@@ -0,0 +1,32 @@
+                />
+            </xpath>
+        </field>
+    </record>
```
We should probably move these inside `commission_settlement_views.xml`.

yostashiro commented on account_commission/views/res_partner.xml:
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="view_partner_form_agent" model="ir.ui.view">
+        <field name="name">res.partner.form.agent</field>
+        <field name="model">res.partner</field>
+        <field name="inherit_id" ref="base.view_partner_form" />
+        <field name="priority" eval="18" />
+        <field name="groups_id" eval="[(4, ref('account.group_account_invoice'))]" />
+        <field name="arch" type="xml">
+            <xpath
+                expr="//page[@name='agent_information']//group/group"
+                position="after"
+            >
+                <group
+                    colspan="4"
+                    string="Settlements"
+                    attrs="{'invisible': [('settlement_ids', '=', [])]}"
+                >
+                    <field name="settlement_ids" nolabel="1">
+                        <tree
+                            decoration-danger="state == 'cancel'"
+                            decoration-success="state == 'invoiced'"
+                        >
+                            <field name="date_from" />
+                            <field name="date_to" />
+                            <field name="total" sum="Settled total" />
+                            <field name="state" />
+                        </tree>
+                    </field>
+                </group>
+            </xpath>
+        </field>
+    </record>
```
Would it be more appropriate to move this record to `commission` and add record rules to make `sale_invoice` settlement records only visible to `account.group_account_invoice`, now that `manual` settlement is allowed?

pedrobaeza replied:
Security rules are already set, so only moving it is needed.

yostashiro commented on commission/readme/CONFIGURE.rst:
```diff
@@ -0,0 +1,38 @@
+For adding commissions:
+
+#. Go to *Invoicing > Configuration > Commission Management > Commission types*.
```
```suggestion
#. Go to *Commissions > Configuration > Commission types*.
```

yostashiro commented on commission/readme/USAGE.rst:
```diff
@@ -0,0 +1,35 @@
+
+For settling the commissions to agents:
+
+#. Go to *Invoicing > Vendors > Commissions > Settle commissions*.
```
```suggestion
#. Go to *Commissions > Settlements > Settle Commissions*.
```

yostashiro commented on commission/readme/USAGE.rst:
```diff
@@ -0,0 +1,35 @@
+   section. Leave it empty for settling all.
+#. Click on "Make settlements" button.
+#. If there are new settlements, they will be shown after this.
+
+For invoicing the settlements (only for external agents):
+
+#. Go to *Invoicing > Vendors > Create commission invoices*.
+#. On the window that appears, you can select following data:
+
+   * Product. It should be a service product for being coherent.
+   * Journal: To be selected between existing purchase journals.
+   * Date: If you want to choose a specific invoice date. You can leave it
+     blank if you prefer.
+   * Settlements: For selecting specific settlements to invoice. You can leave
+     it blank as well for invoicing all the pending settlements.
+
+#. If you want to invoice a specific settlement, you can navigate to it in
+   *Invoicing > Vendors > Settlements*, and click on "Make invoice"
+   button.
```
```suggestion
```

yostashiro commented on account_commission/readme/USAGE.rst:
```diff
@@ -0,0 +1,17 @@
+#. You have a button "Regenerate agents" on the bottom of the page
+   "Invoice Lines" for forcing a recompute of all agents from the partner setup.
+   This is needed for example when you have changed the partner on the
+   invoice having already inserted lines.
```
```suggestion
   invoice having already inserted lines.

For invoicing the settlements (only for external agents):

#. Go to *Invoicing > Commissions > Invoices Settlements* (or *Commissions > Settlements > Settlements*) and open a settlement record.
#. Click "Make invoice" button, and on the window that appears, you can select following data:

   * Product. It should be a service product for being coherent.
   * Journal: To be selected between existing purchase journals.
   * Date: If you want to choose a specific invoice date. You can leave it
     blank if you prefer.
   * Settlements: For selecting specific settlements to invoice. You can leave
     it blank as well for invoicing all the pending settlements.
```

yostashiro commented on account_commission/models/account_move.py:
```diff
@@ -0,0 +1,233 @@
+            raise exceptions.ValidationError(
+                _("You can't cancel an invoice with settled lines"),
+            )
+        self.mapped("line_ids.settlement_id").write({"state": "except_invoice"})
```
I think we need to take care of unlink cases as well.

pedrobaeza replied:
In fact, for unlinking or cancelling, you need to be in draft state, so I have changed the overrided method to `button_draft`.

elvise commented:
@pedrobaeza  I think "Agents" instead of "Commissions" makes more sense

![C2B53692-6466-4A65-BA15-C214FCB00E69](https://user-images.githubusercontent.com/5269577/204856597-49f674a6-f588-4f8a-8234-5c714eaf07fb.jpeg)

yostashiro commented:
@pedrobaeza We would like to expedite this migration due to our project situation.  If you are too busy to update this PR soon, we will try following up with another PR, adding a commit on top of yours.  Would you like us to do so?  In that case, can you please comment on each suggested point with your opinion/judgment so that your ideas will be reflected in our update.

pedrobaeza commented:
Sorry, and thanks for the quality feedback. I'm on it.

pedrobaeza committed:
[IMP] commission*: Add settlement_type to commissions

This way, we can restrict which commissions we can select or be
automatically populate in the different documents.

pedrobaeza commented:
> I think "Agents" instead of "Commissions" makes more sense

Not on the main menu, as that one contains all the commission related things, including the agents.

pedrobaeza commented:
> In my opinion, I think enterprise users may confuse a little bit for home app screen UX. Two apps related with accounting will be displayed in home screen. But just a minor thing.

@AungKoKoLin1997 fixed using this trick: https://github.com/OCA/commission/pull/387/files#diff-0be2abbed382f1580ef0a0f9ee00f07e299ef4d714aeecad3d64b1db2ebefab9R8

pedrobaeza commented:
Everything sorted out or commented. Please tell me if anything more is found.

yostashiro commented on account_commission/models/commission.py:
```diff
@@ -0,0 +1,14 @@
+    invoice_state = fields.Selection(
+        [("open", "Invoice Based"), ("paid", "Payment Based")],
+        string="Invoice Status",
+        required=True,
```
```suggestion
```

yostashiro commented on account_commission/views/commission_views.xml:
```diff
@@ -0,0 +1,13 @@
+        <field name="inherit_id" ref="commission.commission_form" />
+        <field name="arch" type="xml">
+            <xpath expr="//group[@name='commision_params']" position="inside">
+                <field name="invoice_state" />
```
```suggestion
                <field name="invoice_state" attrs="{'invisible': [('settlement_type', '!=', 'sale_invoice')], 'required': [('settlement_type', '=', 'sale_invoice')]}" />
```

pedrobaeza replied:
If no `settlement_type` set (available for all type of commissions), it should be also possible to be set. I'm going to put only the required one, but not to hide it.

yostashiro commented on account_commission/wizards/wizard_invoice.xml:
```diff
@@ -0,0 +1,46 @@
+        <field name="target">new</field>
+        <field name="binding_model_id" ref="model_commission_make_invoice" />
+    </record>
+</odoo>
```
```suggestion
    <menuitem
        id="menu_commission_make_invoices"
        parent="menu_invoicing_commission"
        action="action_commission_make_invoices"
        sequence="20"
    />
</odoo>
```

yostashiro replied:
Somehow the menu item was missing.

pedrobaeza replied:
This one doesn't work through wizard, but it's an action bound to the settlements. The flow is:

- You select one or several settlements.
- Click on Action > Create Commission Invoices.
- The wizard appears.

pedrobaeza replied:
OK, but I see the other flow you want. I'll implement both.

yostashiro commented on account_commission/readme/USAGE.rst:
```diff
@@ -0,0 +1,33 @@
+
+For invoicing the settlements (only for external agents):
+
+#. Go to *Invoicing > Vendors > Create commission invoices*.
```
```suggestion
#. Go to *Invoicing > Commissions > Create commission invoices*.
```

yostashiro commented on account_commission/views/commission_settlement_views.xml:
```diff
@@ -0,0 +1,73 @@
+            >
+                <field
+                    name="invoice_line_id"
+                    attrs="{'column_invisible': [('parent.settlement_type', '!=', 'invoice')]}"
```
```suggestion
                    attrs="{'column_invisible': [('parent.settlement_type', '!=', 'sale_invoice')]}"
```

yostashiro commented on account_commission/views/commission_settlement_views.xml:
```diff
@@ -0,0 +1,73 @@
+                <field
+                    name="invoice_line_id"
+                    attrs="{'invisible': [('parent.settlement_type', '=', 'invoice')]}"
+                />
```
```suggestion
                <field name="invoice_line_id" />
```

yostashiro commented on account_commission/models/account_move.py:
```diff
@@ -0,0 +1,233 @@
+            raise exceptions.ValidationError(
+                _("You can't cancel an invoice with settled lines"),
+            )
+        self.mapped("line_ids.settlement_id").write({"state": "except_invoice"})
```
```suggestion
        self.mapped("line_ids.settlement_id").write({"state": "invoiced"})
```
We should not do this here for button_draft() as it makes the settlement stuck at `except_invoce` state with no way to re-create a vendor bill for the settlement. IMO, no state update should be done on settlement when vendor bill state is changed from 'posted' to 'draft'. Setting `except_invoice` on the settlement should be done in button_cancel() as it did in 14.0. Instead, in button_draft(), we may actually update the settlement to `invoiced` state to handle the cases where vendor bill state is changed from 'cancel' to 'draft'.  Unlink case should still be handled to update the state of the settlement back to `settled`, because, otherwise, the settlement state will be stuck in the `except_invoice` state when 'cancel' invoice is deleted.

pedrobaeza replied:
OK, I have put it back this on cancel method, but added a way to reset a settlement in except state to draft for being re-invoiced.

yostashiro reviewed (commented):
@pedrobaeza Thank you very much for your fast follow-up. I did some functional tests and there are a few more points to take care of.

pedrobaeza commented:
Changes pushed.

yostashiro commented on account_commission/models/account_move.py:
```diff
@@ -0,0 +1,233 @@
+        """Check settled lines and put settlements associated to the invoices in
+        exception.
+        """
+        if any(self.mapped("invoice_line_ids.any_settled")):
+            raise exceptions.ValidationError(
+                _("You can't cancel an invoice with settled lines"),
+            )
```
I was thinking that this check should stay in button_draft() to ensure the consistency.  I think doing this check in cancel step is a remnant of 12.0 in which cancel was the step to take to revert the posted invoice, which has changed in 13.0.  IMO, it should be done in button_draft() now, but it's OK if it's handled in another PR also.

pedrobaeza replied:
Yes, that's was my initial thought, but in fact being a draft means the same as other areas: a draft invoice in sales means that the order is not invoiceable while not cancelled, so this one is a good balance being able to re-invoice the settlement clicking the button.

yostashiro reviewed (approved):
Code review. 👍 Thank you very much!

pedrobaeza commented:
Let's go with this! I will add later the migration scripts for transforming 14.0 datamodel to this one:

/ocabot merge major

mgielissen commented:
No migrations script?

pedrobaeza commented:
There are 2 options:

- You do them and contribute it.
- Wait until I do it, but not now: https://github.com/OCA/commission/pull/387#issuecomment-1337303688. They are not needed for vanilla 15.0 DBs, so not a requirement for merging.

hildickethan commented:
@pedrobaeza From what I see in the migration commit, the migration script would be primarily renaming models and fields. I suppose the biggest part is how to handle the module splitting.
~~Should an Odoo with 14.0 `sale_commission` that migrates to 15.0 have `account_commission` auto-installed too? Since it's part of the module they had, if they don't use that part then it can be uninstalled later.~~ I didn't see that it depends on it, I suppose that's not a problem then

I'm looking to migrate this to 16.0, but I don't really have a good test subject to help with this migration script right now, do you have anything planned for it? 

Edit: Now that the PR is done and I mentioned you again, probably better to reply there https://github.com/OCA/commission/pull/394

pedrobaeza commented:
Yes, basically is to rename some tables, models, columns and fields, and set some values. I will do this on February.
